lib/pure/collections/smart_ptr.nim

```diff
@@ -0,0 +1,152 @@
+#
+#
+#            Nim's Runtime Library
+#        (c) Copyright 2019 Nim contributors
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+
+import allocators
+
+type
+  UniquePtr*[T] = object
+    ## non copyable pointer to object T, exclusive ownership of the object is assumed
+    val: ptr tuple[value: T, allocator: Allocator]
+
+proc `=destroy`*[T](p: var UniquePtr[T]) =
+  mixin `=destroy`
+  if p.val != nil:
+    `=destroy`(p.val[])
+    p.val.allocator.dealloc(p.val.allocator, p.val, sizeof(p.val[]))
+    p.val = nil
+
+proc `=`*[T](dest: var UniquePtr[T], src: UniquePtr[T]) {.error.}
+
+proc `=sink`*[T](dest: var UniquePtr[T], src: UniquePtr[T]) {.inline.} =
+  if dest.val != src.val:
+    if dest.val != nil:
+      dest.val = src.val
+
+proc newUniquePtr*[T](val: sink T): UniquePtr[T] =
+  let a = getSharedAllocator()
+  result.val = cast[type(result.val)](a.alloc(a, sizeof(result.val[])))
+  if AllocatorFlag.ZerosMem notin a.flags:
+    reset(result.val[])
+  result.val.value = val
+  result.val.allocator = a
+
+converter convertUniquePtrToObj*[T](p: UniquePtr[T]): var T {.inline.} =
+  p.val.value
+
+proc isNil*[T](p: UniquePtr[T]): bool {.inline.} =
+  p.val == nil
+
+proc get*[T](p: UniquePtr[T]): var T {.inline.} =
+  when compileOption("boundChecks"):
+    if p.val == nil:
+      raise newException(ValueError, "deferencing nil unique pointer")
+  p.val.value
+
+proc `$`*[T](p: UniquePtr[T]): string {.inline.} =
+  if p.val == nil: "UniquePtr[" & $T & "](nil)"
+  else: "UniquePtr[" & $T & "](" & $p.val.value & ")"
+
+#------------------------------------------------------------------------------
+
+type
+  SharedPtr*[T] = object
+    ## shared ownership reference counting pointer
+    val: ptr tuple[value: T, atomicCounter: int, allocator: Allocator]
+
+proc `=destroy`*[T](p: var SharedPtr[T]) =
+  mixin `=destroy`
+  if p.val != nil:
+    let c = atomicDec(p.val[].atomicCounter)
```

Comment:
You can perform this ``atomicDec`` in an ``else`` when you change the check to ``atomicRead(p.val[].atomicCounter) == 1``. And ``== 0`` would be better still, for this you need to store the value of RC-1.